internal/server/server.go

```diff
@@ -17,209 +17,255 @@ import (
+		stoppableServer, err = startSSEServer(ctx, srv, listenAddr, serveOpts, authContext, serverErrChan)
 
-		err := streamableServer.Start(listenAddr)
-		if err != nil && !errors.Is(err, http.ErrServerClosed) {
-			slog.ErrorContext(ctx, "failed to serve MCP server via HTTP", "error", err)
-		}
-	}()
+	case utils.TransportStreamable:
+		stoppableServer, err = startStreamableHTTPServer(ctx, srv, listenAddr, serveOpts, authContext, serverErrChan)
 
-	slog.InfoContext(ctx, "mcp server via HTTP listening successfully", "server.address", listenAddr)
+	default:
+		return fmt.Errorf("invalid transport type: %s", serveOpts.Transport)
+	}
+	if err != nil {
+		return err
+	}
 
-	return streamableServer
+	// Block until shutdown or an error occurs
+	return waitForShutdown(ctx, stoppableServer, serverErrChan)
 }
 
-func startSSEServer(
+// startServer starts a given MCP in a goroutine.
+func startServer(
 	ctx context.Context,
-	mcpSrv *mcpserver.MCPServer,
 	listenAddr string,
-	serveOpts *ServeOptions,
-) error {
-	authContextFuncWrapper := func(c context.Context, r *http.Request) context.Context {
-		if !serveOpts.OauthEnabled {
-			return authContextFuncNoOauth(
-				c,
-				r,
-				serveOpts.OauthValidateURL,
-				serveOpts.TrentoURL,
-				serveOpts.TrentoUsername,
-				serveOpts.TrentoPassword,
+	server StoppableServer,
+	transportType utils.TransportType,
+	errChan chan<- error,
+) {
+	go func() {
+		slog.InfoContext(ctx, "the MCP server is listening",
+			"server.address", listenAddr,
+			"server.transport", transportType,
+		)
+		if err := server.Start(listenAddr); err != nil && !errors.Is(err, http.ErrServerClosed) {
+			slog.ErrorContext(ctx, fmt.Sprintf("failed to serve MCP server via %s", transportType),
+				"error", err,
 			)
+			errChan <- err
 		}
+	}()
+}
 
-		return authContextFunc(
-			c,
-			r,
-			serveOpts.OauthValidateURL,
-			serveOpts.TrentoURL,
-			serveOpts.TrentoUsername,
-			serveOpts.TrentoPassword,
-		)
-	}
-
-	// Create the server, using custom one to handle mcp auth
-	sseServer := mcpserver.NewSSEServer(mcpSrv, mcpserver.WithSSEContextFunc(authContextFuncWrapper))
-
-	// Run the http server for the mcp in a separate goroutine.
-	slog.InfoContext(ctx, "mcp server via SSE starting", "server.address", listenAddr)
-
-	err := sseServer.Start(listenAddr)
-	if err != nil && !errors.Is(err, http.ErrServerClosed) {
-		slog.ErrorContext(ctx, "failed to serve MCP server via HTTP", "error", err)
-		panic(err)
-	}
-
-	slog.InfoContext(ctx, "mcp server via SSE listening successfully", "server.address", listenAddr)
+// startStreamableHTTPServer initializes and starts a custom streamable HTTP server.
+func startStreamableHTTPServer(
+	ctx context.Context,
+	mcpSrv *mcpserver.MCPServer,
+	listenAddr string,
+	serveOpts *ServeOptions,
+	authContext authContextWrapperFn,
+	errChan chan<- error,
+) (*CustomStreamableHTTPServer, error) {
+	streamableServer := NewCustomStreamableHTTPServer(mcpSrv, "/mcp", authContext, serveOpts)
```

Comment:
For now, hardcoded to `/mcp`   (the SSE endpoint also hardcodes it to `/sse`, but before changing that, I want to see the implications, especially when running along with the `mcpo` proxy.